File: lib/Arrays-LC/spiral_matrix.rb
# Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

def spiral_matrix(matrix)
    return '[]' if matrix.empty?

    m = matrix.dup
    result = []

    until m.size < 1
        result << m.shift
        m = m.transpose.reverse
    end

    #  for working with larger matrices
    # until m.size < 1
    #     result << m.shift
    #     result << m.map(&:pop)
    #     result << (m.pop || []).reverse
    #     result << m.map(&:shift).reverse
    # end

    result.flatten
end

# testing

matrix = [
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]

puts "spiral_matrix = #{spiral_matrix(matrix)}"

matrix = [[6,9,7]]

puts "spiral_matrix = #{spiral_matrix(matrix)}"

File: lib/JS/chapter_1/spiralMatrix.js
// Given a matrix of m x n elements(m rows, n columns), return all elements of the matrix in spiral order.

const spiralMatrix = matrix => {
    if (matrix.length === 0) return []

    m = matrix.slice(0)
    result = []

    while(m.length > 0){
        result.push(m.shift())
        if(m.length === 0) break;
        result.push(m.map( el => el.pop()))
        result.push((m.pop() || []).reverse())
        result.push(m.map( el => el.shift()).reverse())
    }

    return result

}

// testing

matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]
console.log(`Spiral matrix for ${matrix}: ${spiralMatrix(matrix)}`);

matrix = []
console.log(`Spiral matrix for ${matrix}: ${spiralMatrix(matrix)}`);

matrix = [[6, 9, 7]]
console.log(`Spiral matrix for ${matrix}: ${spiralMatrix(matrix)}`);